diff --git a/extern/KaHIP/lib/io/graph_io.cpp b/extern/KaHIP/lib/io/graph_io.cpp
index 8b94911..e23ef98 100644
--- a/extern/KaHIP/lib/io/graph_io.cpp
+++ b/extern/KaHIP/lib/io/graph_io.cpp
@@ -142,11 +142,11 @@ int graph_io::readGraphWeighted(graph_access & G, std::string filename) {
                 if( read_nw ) {
                         ss >> weight;
                         total_nodeweight += weight;
-                        if( total_nodeweight > (long long) std::numeric_limits<NodeWeight>::max()) {
-                                std::cerr <<  "The sum of the node weights is too large (it exceeds the node weight type)."  << std::endl;
-                                std::cerr <<  "Currently not supported. Please scale your node weights."  << std::endl;
-                                exit(0);
-                        }
+                        // if( total_nodeweight > (long long) std::numeric_limits<NodeWeight>::max()) {
+                        //         std::cerr <<  "The sum of the node weights is too large (it exceeds the node weight type)."  << std::endl;
+                        //         std::cerr <<  "Currently not supported. Please scale your node weights."  << std::endl;
+                        //         exit(0);
+                        // }
                 }
                 G.setNodeWeight(node, weight);
 
@@ -269,11 +269,11 @@ int graph_io::readGraphWeightedFast(graph_access & G, std::string filename) {
                 if( read_nw ) {
                         weight = naive(line, line_ptr);
                         total_nodeweight += weight;
-                        if( total_nodeweight > (long long) std::numeric_limits<NodeWeight>::max()) {
-                                std::cerr <<  "The sum of the node weights is too large (it exceeds the node weight type)."  << std::endl;
-                                std::cerr <<  "Currently not supported. Please scale your node weights."  << std::endl;
-                                exit(0);
-                        }
+                        // if( total_nodeweight > (long long) std::numeric_limits<NodeWeight>::max()) {
+                        //         std::cerr <<  "The sum of the node weights is too large (it exceeds the node weight type)."  << std::endl;
+                        //         std::cerr <<  "Currently not supported. Please scale your node weights."  << std::endl;
+                        //         exit(0);
+                        // }
                 }
                 G.setNodeWeight(node, weight);
 
diff --git a/extern/KaHIP/lib/parallel_mh_clustering/population_clustering.cpp b/extern/KaHIP/lib/parallel_mh_clustering/population_clustering.cpp
index cd358b4..1759eb7 100644
--- a/extern/KaHIP/lib/parallel_mh_clustering/population_clustering.cpp
+++ b/extern/KaHIP/lib/parallel_mh_clustering/population_clustering.cpp
@@ -35,6 +35,8 @@ population_clustering::population_clustering( MPI_Comm communicator, const Parti
         m_time_stamp         = 0;
         m_communicator       = communicator;
         m_global_timer.restart();
+        best_objective = -1;
+        m_best_objective_time = 0.0;
 }
 
 population_clustering::~population_clustering() {
@@ -95,6 +97,12 @@ void population_clustering::createIndividuum(const PartitionConfig & config,
 }
 
 void population_clustering::insert(graph_access & G, Individuum & ind) {
+        if( ind.objective > best_objective ) {
+                m_filebuffer_string <<  m_global_timer.elapsed() <<  " " <<  ind.objective <<  std::endl;
+                m_time_stamp++;
+                best_objective = ind.objective;
+                m_best_objective_time = m_global_timer.elapsed();
+        }
 
         m_no_partition_calls++;
         if(m_internal_population_clustering.size() < m_population_clustering_size) {
diff --git a/extern/KaHIP/lib/parallel_mh_clustering/population_clustering.h b/extern/KaHIP/lib/parallel_mh_clustering/population_clustering.h
index 0feb3a1..2897dd3 100644
--- a/extern/KaHIP/lib/parallel_mh_clustering/population_clustering.h
+++ b/extern/KaHIP/lib/parallel_mh_clustering/population_clustering.h
@@ -117,6 +117,8 @@ class population_clustering {
 
                 void write_log(std::string & filename);
 
+                double get_best_objective_time() { return m_best_objective_time; }
+
                 /* updates old clustering with a new clustering of a possibly contracted graph.
                  * new_coarse_clustering must be of size max(clustering) + 1, i.e. must have an
                  * entry for every cluster in the old clustering. */
@@ -325,6 +327,7 @@ class population_clustering {
 
         private:
 
+                double                  m_best_objective_time;
                 unsigned                m_no_partition_calls;
                 unsigned 		m_population_clustering_size;
                 std::vector<Individuum> m_internal_population_clustering;
@@ -335,6 +338,7 @@ class population_clustering {
                 int m_num_NCs_computed;
                 int m_num_ENCs;
                 int m_time_stamp;
+                double best_objective;
 
                 MPI_Comm m_communicator;
 
diff --git a/lib/parallel_mh_clustering/parallel_mh_async_clustering.cpp b/lib/parallel_mh_clustering/parallel_mh_async_clustering.cpp
index 28901c4..1bcc917 100644
--- a/lib/parallel_mh_clustering/parallel_mh_async_clustering.cpp
+++ b/lib/parallel_mh_clustering/parallel_mh_async_clustering.cpp
@@ -143,6 +143,8 @@ double parallel_mh_async_clustering::collect_best_partitioning(graph_access & G,
         double best_local_objective   = max_objective;
         double best_local_objective_m = max_objective;
         double best_global_objective  = 0;
+        double best_local_time        = m_island->get_best_objective_time();
+        double best_global_time       = 0;
 
         PartitionID* best_local_map = new PartitionID[G.number_of_nodes()];
 
@@ -157,12 +159,22 @@ double parallel_mh_async_clustering::collect_best_partitioning(graph_access & G,
         int g_bcaster = 0;
 
         MPI_Allreduce(&bcaster, &g_bcaster, 1, MPI_INT, MPI_MIN, m_communicator);
+        
+        if( m_rank == g_bcaster ) {
+                best_global_time = best_local_time;
+        }
+
+        MPI_Bcast(&best_global_time, 1, MPI_DOUBLE, g_bcaster, m_communicator);
         MPI_Bcast(best_local_map, G.number_of_nodes(), MPI_INT, g_bcaster, m_communicator);
 
         forall_nodes(G, node) {
                 G.setPartitionIndex(node, best_local_map[node]);
         } endfor
 
+        if( m_rank == ROOT) {
+                std::cout << "Best solution found after " << best_global_time << " s" << std::endl;
+        }
+
         delete[] best_local_map;
 
         return best_global_objective;
diff --git a/lib/parallel_mh_clustering/population_clustering.cpp b/lib/parallel_mh_clustering/population_clustering.cpp
index 5d815cb..a891f3d 100644
--- a/lib/parallel_mh_clustering/population_clustering.cpp
+++ b/lib/parallel_mh_clustering/population_clustering.cpp
@@ -36,6 +36,7 @@ population_clustering::population_clustering( MPI_Comm communicator, const Parti
         m_communicator       = communicator;
         m_global_timer.restart();
         best_objective = -1;
+        m_best_objective_time = 0.0;
 }
 
 population_clustering::~population_clustering() {
@@ -106,6 +107,7 @@ void population_clustering::insert(graph_access & G, Individuum & ind) {
                 m_filebuffer_string <<  m_global_timer.elapsed() <<  " " <<  ind.objective <<  std::endl;
                 m_time_stamp++;
                 best_objective = ind.objective;
+                m_best_objective_time = m_global_timer.elapsed();
         }
 
         m_no_partition_calls++;
diff --git a/lib/parallel_mh_clustering/population_clustering.h b/lib/parallel_mh_clustering/population_clustering.h
index 45e5ae1..18ba509 100644
--- a/lib/parallel_mh_clustering/population_clustering.h
+++ b/lib/parallel_mh_clustering/population_clustering.h
@@ -116,6 +116,8 @@ class population_clustering {
 
                 void write_log(std::string & filename);
 
+                double get_best_objective_time() { return m_best_objective_time; }
+
                 /* updates old clustering with a new clustering of a possibly contracted graph.
                  * new_coarse_clustering must be of size max(clustering) + 1, i.e. must have an
                  * entry for every cluster in the old clustering. */
@@ -328,6 +330,7 @@ class population_clustering {
 
         private:
 
+                double                  m_best_objective_time;
                 unsigned                m_no_partition_calls;
                 unsigned 		m_population_clustering_size;
                 std::vector<Individuum> m_internal_population_clustering;
