diff --git a/DGI.py b/DGI.py
index 3df6ef4..28939fc 100644
--- a/DGI.py
+++ b/DGI.py
@@ -66,9 +66,9 @@ class DeepGraphInfomax(torch.nn.Module):
         self.reset_parameters()
         self.K = args.K
         self.cluster_temp = args.clustertemp
-        self.init = torch.rand(self.K,hidden_channels)
+        self.register_buffer("init", torch.rand(self.K, hidden_channels))
         self.cluster = cluster
-        
+
     def reset_parameters(self):
         reset(self.encoder)
         reset(self.summary)
@@ -117,14 +117,14 @@ class DeepGraphInfomax(torch.nn.Module):
         return -torch.log(self.discriminate(pos_z,self.summary(mu),sigmoid=True) + EPS).mean()
 
     def modularity(self, mu, r, embeds, dist, bin_adj, mod, args):
-        bin_adj_nodiag = bin_adj * (torch.ones(bin_adj.shape[0], bin_adj.shape[0]) - torch.eye(bin_adj.shape[0]))
+        bin_adj_nodiag = bin_adj * (torch.ones(bin_adj.shape[0], bin_adj.shape[0], device=bin_adj.device) - torch.eye(bin_adj.shape[0], device=bin_adj.device))
         return (1. / bin_adj_nodiag.sum()) * (r.t() @ mod @ r).trace()
 
     def test(self, train_z, train_y, test_z, test_y, solver='lbfgs',
              multi_class='auto', *args, **kwargs):
         r"""Evaluates latent space quality via a logistic regression downstream
         task."""
-        clf = LogisticRegression(solver=solver, multi_class=multi_class, *args,
+        clf = LogisticRegression(solver=solver, *args,
                                  **kwargs).fit(train_z.detach().cpu().numpy(),
                                                train_y.detach().cpu().numpy())
         return clf.score(test_z.detach().cpu().numpy(),
diff --git a/evaluation.py b/evaluation.py
index a89e76b..9f56544 100644
--- a/evaluation.py
+++ b/evaluation.py
@@ -78,7 +78,7 @@ def matched(true,pred):
     cm = Metrics.confusion_matrix(true,pred,labels=np.arange(0,max_idx+1))
     shifted_mat = np.zeros((cm.shape[0]*2,cm.shape[0] * 2))
     shifted_mat[:cm.shape[0],cm.shape[0]:] = cm
-    g = nx.from_numpy_matrix(shifted_mat)
+    g = nx.from_numpy_array(shifted_mat)
     match = nx.max_weight_matching(g)
     unmatched = set(np.arange(0,cm.shape[0]))
     label_map = {}
